engineering-plans/0004-subgraph-grafting.md

```diff
@@ -0,0 +1,143 @@
+# PLAN-0004: Subgraph Grafting
+
+<dl>
+  <dt>Author</dt>
+  <dd>David Lutterkort</dd>
+
+  <dt>Implements</dt>
+  <dd>No RFC</dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="URL">URL</a></dd>
+
+  <dt>Date of submission</dt>
+  <dd>2020-03-17</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Contents
+
+<!-- toc -->
+
+## Summary
+
+This feature makes it possible to resume indexing of a failed subgraph by
+creating a new subgraph that uses the failed subgraph's data up to a block
+that the subgraph writer deems reliable, and continue indexing using the
+new subgraph's mappings. We call this operation *grafting* subgraphs.
+
+## Subgraph manifest changes
+
+The user-facing control for this feature will be a new optional
+`deployment` field in the subgraph's manifest. It can take one of three
+forms, and defaults to `create`:
+
+    ---
+    description: My really very good subgraph
+    # Option 1: (default) The deployment starts with no data in the database
+    deployment:
+      mode: create
+    # Option 2: (copy) The deployment starts with a copy of the data from
+    # the given subgraph at the given block
+    deployment:
+      mode: copy
+      subgraph: Qm...
+      block: 123456
+    # Option 3: (reuse) The deployment reuses the data of an existing subgraph
+    # in place. This mode implicitly deletes the source deployment, and any
+    # subgraph version using the source deployment will be switched to use the
+    # new deployment
```

Comment:
At least on the hosted service, this could mean messing with existing deployments that other people have deployed copies of under different subgraph names and may be relying on. I don't think we want that.

Reply:
Having said that, we could allow it for subgraph deployments that are exclusively used by the subgraph name we're operating against. What do you think? (It may be worth adding a note about this.)

Reply:
The more I think about `reuse`, the more I think we should not do it since it has all kinds of downsides and complications, and the main advantage is to save time on initial setup. We should definitely not do it unless we run into a situation where `copy` is not fast enough, and even then we might not want to allow it in shared setups like hosted.

Reply:
I'm on board with that.

Reply:
I don't think that having modes and deploy time errors is a good way to surface the API in the first place. Semantically, "deploy mode" has no impact on the result of the expected results of a query for a particular block. Re-use, copy, and the like are internal implementation details of graph-node itself, and what decisions are made here may be resulting from external state beyond the subgraph (like whether or not an indexer has both the old and new version of the subgraph deployed). This whole section should be changed to reflect how grafting works semantically and not operationally.

Reply:
I view grafting as more than just an implementation detail, especially `reuse` as it leads to visibly different behavior (one subgraph disappears, another one appears)

But even with `copy`, deploy mode influences what data you ultimately put in your subgraph, e.g., because grafting onto a different block will in general lead to different data.

Having said that, how would you rewrite this to describe it at a more semantic level?

Reply:
In my mind a graft is a completely new subgraph, really exactly like a fork. It semantically executes one function up to a point, then another function from then on. Whether or not the results of that execution are cached and how is an implementation detail. (Eg: An index node that re-indexed from scratch for each query while wildly inefficient would not be semantically incorrect). Where you get the data from the previous blocks functions executing is a question of how the cache is implemented.

The concept of one subgraph disappearing and another appearing has no place in that mental model.

If anything, the graft is more like a data source than anything else.
Eg:
```toml
dataSources:
    - kind: fork
       subgraph: '0x0'
    - kind: ethereum/contract
       startBlock: 0 <- the block that the new function starts executing on
```
(The particulars of where this information go in subgraph.yaml or how this concept is open to debate)

The extent of the semantics that need to be described are "execute one function up to here, then a different function from then on" is easier to understand than what is written, and more powerful. The old subgraph still exists, and may or may not still be queried depending on whether it is deployed.

This model may enable a kind of backward compatible rollout of subgraph updates, or updates to live subgraphs that occur from a future block, as well as the given use case of rapidly deploying fixes.

I feel though I'm on a very different page then the rest of the team here. In the name of forward progress, I'm willing to just give up pressing this issue if that would be more useful.

Reply:
I think the disconnect is in viewing the subgraph data as precious in some way: other than for the fact that reindexing takes time, that data is not precious in the same way that data in other applications is - all subgraph data is derivative and not original, and can be reconstructed. Grafting is an attempt at making the time it takes to reindex/address a problem shorter.

I don't think it is a good tool to express the evolution of logic of subgraph mappings, since it's much harder to see how behavior evolved with grafts than if that was expressed explicitly in the mappings; i.e., with grafting, you have to compare your subgraph code at commits X and Y (or at different tags or branches) to see how logic has changed whereas if that's done in the mapping code you'd be looking at an if statement somewhere.